torch/_decomp/decompositions.py

```diff
@@ -2266,7 +2266,7 @@ def should_fold(tensor1: torch.Tensor, dim_tensor2: int) -> bool:
         return False
 
 
-@torch.ops.aten.matmul.default.py_impl(DispatchKey.CompositeImplicitAutograd)
+@register_decomposition(aten.matmul.default, type="pre_autograd")
```

Comment:
It turns out that I cannot conditionally register the CompositeImplicitAutograd py_impl for aten.matmul in AOTAutograd as we have discussed before. coz faketensor requires this py_impl to be always ON. 

I see there are two kinds of pre_autograd decompositions. One that can be opted-in by backends, to avoid the manual definition of backward decomp, e.g. aten.upsample_bilinear2d. One that is part of the our system, e.g. aten.matmul and aten.broadcast_tensor. 

In the same line of thinking, most of ours existing meta functions are an intrinsic part of the system, and they would should be registered when "import torch" happens. However, we do want to accept user supplied meta functions. These can be passed in through aotautograd's decompositions argument, and registered dynamically.

Reply:
cc @ezyang, @Chillee

Reply:
In the meeting, we were saying all decomps needs to be registered via aotautograd. 
This means even for decomps that needs to be always on, e.g. 'pre_autograd decomp' for aten.matmul, it need to go through aotautograd. This doesn't look like the correct design. 

My proposal is that, there are two entries for registering decomps
1. one that happens at "import torch". This is for critical decomps that the system cannot live without, e.g. meta functions. 
2. one that happens at aotautograd, which allows backend's customization.

Reply:
yeah ok we'll have to talk about this